libvirt/tests/src/virtiofs/virtiofs.py

```
@@ -289,46 +301,168 @@ def check_filesystem_hotplug_with_mem_setup():
         else:
             open_files_max = None
 
-        # Define filesystem device xml
-        for index in range(fs_num):
-            driver = {'type': driver_type, 'queue': queue_size}
-            source_dir = os.path.join('/var/tmp/', str(dir_prefix) + str(index))
+        def _get_fs_dev_and_source_dir(fs_index, socket_index):
+            """
+            Returns the device XML for both internally or externally launched
+            virtiofsd.
+
+            :param fs_index: the index of the filesystem
+            :param socket_index: the index of the socket for externally launched
+                                 virtiofsd; this parameter is ignored for internally
+                                 launched virtiofsd and must be different for each
+                                 VM accessing the filesystem
+            :return tuple: (fs_dev, source_dir) - the device xml but also the source
+                           directory which is not part of the XML so the test can
+                           launch the instance for it
+            """
+            driver = {'type': driver_type}
+            if queue_size != "":
+                driver['queue'] = queue_size
+            source_dir = os.path.join('/var/tmp/', str(dir_prefix) + str(fs_index))
             logging.debug(source_dir)
-            not os.path.isdir(source_dir) and os.mkdir(source_dir)
-            target_dir = dir_prefix + str(index)
-            source = {'socket': source_socket}
+            if not os.path.isdir(source_dir):
+                if fs_index == 0 and omit_dir_at_first:
+                    pass
+                else:
+                    os.mkdir(source_dir)
```

here can we change it to:
if not (fs_index == 0 and omit_dir_at_first):
    os.mkdir(source_dir)

---

I see the advantage of having the default behavior in the first branch. I thought with the negation it was harder to understand though. In any case `if not (omit_dir_at_first and fs_index == 0)` is not really too difficult so I changed it to that. Thanks

Regarding line 485, I think what you are saying is that it could make sense to remove the `source_dir` explicitly if it existed. However, I think that's unlikely because it's a test-specific name and the directory must be removed by previous test cases if they created it already during the cleanup which I reviewed to make it more stable: the cleanup now reports an error if the directory wasn't removed. Finally, if the directory existed the test should fail, so there's little risk of getting an invalid test result. Therefore, I don't want to change the directory if it existed because it would make the test code more complicated and it's not really necessary.